File: 02_ProgrammingFundamentals/_09_MapsLambdaAndStreamAPI/_03_MoreExercises/_05_DragonArmy.java
package _02_ProgrammingFundamentals._09_MapsLambdaAndStreamAPI._03_MoreExercises;

import java.util.*;
import java.util.stream.Collectors;

public class _05_DragonArmy {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        // Read input value
        int inputLines = Integer.parseInt(scan.nextLine());

        // Add variable
        Map<String, Map<String, List<Double>>> database = new LinkedHashMap<>();

        // Create database for dragons
        for (int i = 0; i < inputLines; i++) {
            String[] information = scan.nextLine().split(" ");

            String type = information[0];
            String name = information[1];
            double damage = information[2].equals("null") ? 45 : Double.parseDouble(information[2]);
            double health = information[3].equals("null") ? 250 : Double.parseDouble(information[3]);
            double armor = information[4].equals("null") ? 10 : Double.parseDouble(information[4]);

            database.putIfAbsent(type, new TreeMap<>());
            List<Double> stats = new ArrayList<>();
            Collections.addAll(stats, damage, health, armor);
            database.get(type).put(name, stats);
        }

        // Print results
        database.entrySet().forEach(type -> {
            String typeOf = type.getKey();
            
            double dmgAverage = type.getValue().values().stream().map(e -> e.get(0)).mapToDouble(Double::valueOf).sum() / type.getValue().size();
            double healthAverage = type.getValue().values().stream().map(e -> e.get(1)).mapToDouble(Double::valueOf).sum() / type.getValue().size();
            double armorAverage = type.getValue().values().stream().map(e -> e.get(2)).mapToDouble(Double::valueOf).sum() / type.getValue().size();

            String firstLine = String.format("%s::(%.2f/%.2f/%.2f)", typeOf, dmgAverage, healthAverage, armorAverage);
            String secondLine = type.getValue().entrySet().stream().map(e -> String.format("-%s -> damage: %.0f, health: %.0f, armor: %.0f", e.getKey(), e.getValue().get(0), e.getValue().get(1), e.getValue().get(2))).collect(Collectors.joining("\n"));

            System.out.printf("%s%n%s%n", firstLine, secondLine);
        });
        
        scan.close();
    }
}

File: 24_C#Fundamentals/_07_AssociativeArrays/_03_MoreExercises/_05_DragonArmy/_05_DragonArmy.cs
using System;

public class _05_DragonArmy
{
    public static void Main(string[] args)
    {
        // Read input value
        int inputLines = int.Parse(Console.ReadLine());

        // Add variable
        Dictionary<string, SortedDictionary<string, List<double>>> database = new Dictionary<string, SortedDictionary<string, List<double>>>();

        // Create database for dragons
        for (int i = 0; i < inputLines; i++)
        {
            string[] information = Console.ReadLine().Split(" ");

            string type = information[0];
            string name = information[1];
            double damage = information[2].Equals("null") ? 45 : double.Parse(information[2]);
            double health = information[3].Equals("null") ? 250 : double.Parse(information[3]);
            double armor = information[4].Equals("null") ? 10 : double.Parse(information[4]);

            if (!database.ContainsKey(type))
            {
                database[type] = new SortedDictionary<string, List<double>>();
            }

            database[type][name] = new List<double> { damage, health, armor };
        }

        // Print results
        foreach (var db in database)
        {
            string type = db.Key;
            SortedDictionary<string, List<double>> dragons = db.Value;

            double avgDamage = dragons.Values.Select(x => x[0]).Average();
            double avgHealth = dragons.Values.Select(x => x[1]).Average();
            double avgArmor = dragons.Values.Select(x => x[2]).Average();

            Console.WriteLine($"{type}::({avgDamage:F2}/{avgHealth:F2}/{avgArmor:F2})");
            foreach (var dragon in dragons)
            {
                string name = dragon.Key;
                double dmg = dragon.Value[0];
                double hp = dragon.Value[1];
                double armor = dragon.Value[2];

                Console.WriteLine($"-{name} -> damage: {dmg:0}, health: {hp:0}, armor: {armor:0}");
            }
        }
    }
}
